File: Algorithms/Math/java/NumberOfDays.java
/**
 * This is a program to count the number of days between two dates. The two
 * dates are given as strings, their format is YYYY-MM-DD as shown in the
 * examples.
 * 
 * Idea of solution: use the juilian days transformation to calculate the number
 * of days on each input and count the difference.
 * 
 * Time Complexity: O(1), Space Complexity: O(1)
 */
public class NumberOfDays {
    /**
     * 
     * @param date1: the first date in YYYY-MM-DD format.
     * @param date2: the second date in YYYY-MM-DD format.
     * @return the difference of days between these two dates.
     */
    public int daysBetweenDates(String date1, String date2) {
        return Math.abs(getDays(date1) - getDays(date2));
    }

    /**
     * 
     * @param date: the data in YYYY-MM-DD format.
     * @return the total days start from Julian day.
     */
    private int getDays(String date) {
        String[] splitDate = date.split("-");
        int y = Integer.valueOf(splitDate[0]), m = Integer.valueOf(splitDate[1]), d = Integer.valueOf(splitDate[2]);
        if (m == 1 || m == 2) {
            m += 12;
            y -= 1;
        }
        return 365 * y + y / 4 - y / 100 + y / 400 + d + (153 * m + 8) / 5;
    }

    public static void main(String[] args) {
        String date1 = "2019-06-29", date2 = "2019-06-30";
        int res = new NumberOfDays().daysBetweenDates(date1, date2);
        System.out.println(res);
    }
}

File: Algorithms/Math/python/number_of_days.py
class Solution:
    def daysBetweenDates(self, date1: str, date2: str) -> int:
        """
        Args:
            date1: the first date in YYYY-MM-DD format.
            date2: the second date in YYYY-MM-DD format.

        Returns:
            return the difference of days between these two dates.

        Example:
            Input: date1 = "2019-06-29", date2 = "2019-06-30"
            Output: 1

            Input: date1 = "2020-01-15", date2 = "2019-12-31"
            Output: 15

        Idea of solution:
            Use the juilian day transformation to calculate the number of days on each input and count the difference.
        
        Time complexity: O(1)
        Space complexity: O(1)
        """
        def get_days(date):
            y, m, d = map(int, date.split("-"))
            if m == 1 or m == 2:
                y -= 1
                m += 12
            return 365 * y + y // 4 + y // 400 - y // 100 + d + (153 * m + 8) // 5
        return abs(get_days(date2) - get_days(date1))